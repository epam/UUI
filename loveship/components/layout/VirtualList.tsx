import { withMods } from '@epam/uui-core';
import { VirtualList as uuiVirtualList, VirtualListProps } from '@epam/uui-components';
import css from './VirtualList.scss';

export interface VirtualListMods {}

function applyVirtualListMods(mods: VirtualListMods) {
    return [css.root];
}

export const VirtualList = withMods<VirtualListProps, VirtualListMods>(uuiVirtualList, applyVirtualListMods);
