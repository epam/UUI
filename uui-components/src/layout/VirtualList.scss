.wrapper {
    flex: 1 1 0;
    align-self: stretch;
    display: flex;
    position: relative;
    width: 100%;
}

.body {
    flex: 1 1 0;
    align-self: stretch;
    display: flex;
}

.container-1 {
    flex: 1 1 0;
    align-self: stretch;
    position: relative;
    overflow: hidden;
    overscroll-behavior-x: none;
}

.container-2 {
    flex: 1 1 0;
    align-self: stretch;
    overflow: hidden;
}

.container-3 {
    display: flex;
    flex-direction: column;
    margin-top: 0;
}

:global(.uui-scroll-shadow-top) {
    position: absolute;
    pointer-events: none;
    left: 0;
    right: 0;
    top: 0;
}

:global(.uui-scroll-shadow-bottom) {
    position: absolute;
    pointer-events: none;
    left: 0;
    right: 0;
    bottom: 0;
}
