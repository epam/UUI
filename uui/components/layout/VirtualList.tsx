import * as React from 'react';
import {
    IHasCX, IEditable, VirtualListState, IHasRawProps, useVirtualList, useScrollShadows, cx, uuiMarkers, IHasChildren, UseVirtualListProps,
} from '@epam/uui-core';
import { ScrollbarsApi } from '@epam/uui-components';
import { ScrollBars } from './ScrollBars';
import css from './VirtualList.module.scss';
import { Blocker } from './Blocker';
import { HTMLAttributes } from 'react';

export interface VirtualListRenderRowsParams<ListContainer extends HTMLElement = any> {
    listContainerRef: React.MutableRefObject<ListContainer>;
    estimatedHeight: number;
    offsetY: number;
    scrollShadows: {
        verticalTop: boolean;
        verticalBottom: boolean;
        horizontalLeft: boolean;
        horizontalRight: boolean;
    };
}
type VirtualListRenderRows<List extends HTMLElement = any> = {
    rows?: React.ReactNode[];
    renderRows: (config: VirtualListRenderRowsParams<List>) => React.ReactNode;
} | {
    rows: React.ReactNode[];
    renderRows?: (config: VirtualListRenderRowsParams<List>) => React.ReactNode;
};

interface BaseVirtualListProps
    extends IHasCX,
    IEditable<VirtualListState>,
    IHasRawProps<HTMLAttributes<HTMLDivElement>>, Pick<UseVirtualListProps, 'rowsCount' | 'rowsSelector' | 'onScroll'> {
    /** HTML role attribute to place on list container */
    role?: React.HTMLAttributes<HTMLDivElement>['role'];
    /** Pass true, to enable Blocker while list loading */
    isLoading?: boolean;
    /**
     * Render callback for virtual list loading blocker
     * If omitted, default UUI `Blocker` component will be rendered
     */
    renderBlocker?: (props: { isLoading: boolean }) => React.ReactNode;
}

export type VirtualListProps<List extends HTMLElement = any> = BaseVirtualListProps & VirtualListRenderRows<List>;

export const VirtualList = React.forwardRef<ScrollbarsApi, VirtualListProps>((props, ref) => {
    const {
        listContainerRef, offsetY, handleScroll, estimatedHeight, scrollContainerRef,
    } = useVirtualList({
        value: props.value,
        onValueChange: props.onValueChange,
        onScroll: props.onScroll,
        rowsCount: props.rowsCount,
        rowsSelector: props.rowsSelector,
    });

    React.useImperativeHandle(ref, () => scrollContainerRef.current, [scrollContainerRef.current]);

    const scrollShadows = useScrollShadows({ root: scrollContainerRef.current });

    const renderRows = () =>
        props.renderRows?.({
            listContainerRef, estimatedHeight, offsetY, scrollShadows,
        }) || (
            <div className={ css.listContainer } style={ { minHeight: `${estimatedHeight}px` } }>
                <div ref={ listContainerRef } role={ props.role } style={ { marginTop: offsetY } }>
                    {props.rows}
                </div>
            </div>
        );

    const renderView = ({ style }: any) => (
        <VirtualListView
            isLoading={ props.isLoading }
            style={ style }
            rawProps={ props.rawProps }
            renderBlocker={ props.renderBlocker }
        />
    );

    const scrollBarsRef = React.useCallback((scrollbars: ScrollbarsApi) => {
        if (!scrollbars?.container?.firstChild) return;
        scrollContainerRef.current = scrollbars.container.firstChild as HTMLDivElement;
    }, []);
    return (
        <ScrollBars
            cx={ cx(css.scrollContainer, props.cx, {
                [uuiMarkers.scrolledLeft]: scrollShadows.horizontalLeft,
                [uuiMarkers.scrolledRight]: scrollShadows.horizontalRight,
                [uuiMarkers.scrolledTop]: scrollShadows.verticalTop,
                [uuiMarkers.scrolledBottom]: scrollShadows.verticalBottom,
            }) }
            onScroll={ handleScroll }
            renderView={ renderView }
            ref={ scrollBarsRef }
        >
            {renderRows()}
        </ScrollBars>
    );
});

interface VirtualListViewProps extends IHasRawProps<HTMLAttributes<HTMLDivElement>>, IHasChildren, Pick<BaseVirtualListProps, 'renderBlocker'> {
    style?: React.CSSProperties;
    isLoading: boolean;
}

const VirtualListView = React.forwardRef<HTMLDivElement, VirtualListViewProps>((props, ref) => {
    return (
        <>
            <div
                { ...props.rawProps }
                style={ {
                    ...props.style,
                    position: 'relative',
                    flex: '1 1 auto',
                    display: 'flex',
                    flexDirection: 'column',
                    overflow: props.isLoading ? 'hidden' : 'scroll',
                    marginRight: props.isLoading ? 0 : props.style.marginRight,
                    marginBottom: props.isLoading ? 0 : props.style.marginBottom,
                } }
                ref={ ref }
            >
                { props.children }
            </div>
            { props.renderBlocker ? props.renderBlocker({ isLoading: props.isLoading }) : <Blocker isEnabled={ props.isLoading } /> }
        </>
    );
});
