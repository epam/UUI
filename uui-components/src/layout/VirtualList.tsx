import * as React from 'react';
import * as ReactDOM from 'react-dom';
import * as css from './VirtualList.scss';
import { IHasCX, IEditable, VirtualListState, cx, IHasRawProps } from '@epam/uui';
import ScrollBars, * as CustomScrollBars from 'react-custom-scrollbars-2';

export interface VirtualListProps extends IHasCX, IEditable<VirtualListState>, IHasRawProps<HTMLDivElement> {
    rows: React.ReactNode[];
    rowsCount?: number;
    focusedIndex?: number;
    role?:  React.HTMLAttributes<HTMLDivElement>['role'];
    onScroll?(value: CustomScrollBars.positionValues): void;
}

export class VirtualList extends React.Component<VirtualListProps, {}> {
    scrollBars: ScrollBars | null;
    container1: Element;
    container2: HTMLElement;
    container3: Element;
    topShadow: HTMLElement | null;
    bottomShadow: HTMLElement | null;
    blockAlign = 20;
    rowHeights: number[] = [];
    rowOffsets: number[] = [];
    estimatedHeight: number = null;
    scrollValues: CustomScrollBars.positionValues = { scrollTop: 0 } as any;

    handleUpdateScroll = () => {
        this.scrollValues = this.scrollBars.getValues();
        if (this.props.onScroll) {
            this.props.onScroll(this.scrollValues);
        }
        let topIndex = 0;
        let scrollTop = this.scrollValues.scrollTop;
        while (topIndex < this.props.rowsCount
            && this.rowOffsets[Math.min(topIndex + this.blockAlign, this.props.rowsCount)] < scrollTop
        ) {
            topIndex += this.blockAlign;
        }

        let bottomIndex = topIndex;
        let scrollBottom = scrollTop + this.scrollValues.clientHeight;
        while (bottomIndex < this.props.rowsCount
            && this.rowOffsets[Math.min(bottomIndex, this.props.rowsCount)] < scrollBottom
        ) {
            bottomIndex++;
        }

        let minVisibleCount = (bottomIndex - topIndex);
        let recommendedVisibleCount = minVisibleCount + this.blockAlign * 2;

        const value = this.props.value;
        if (topIndex != value.topIndex || minVisibleCount > value.visibleCount) {
            this.props.onValueChange({ ...this.props.value, topIndex: topIndex, visibleCount: recommendedVisibleCount });
        }

        // Update shadows visibility
        let showBottomShadow = this.scrollValues.scrollTop == 0
            ? this.estimatedHeight > this.scrollValues.clientHeight
            : (this.scrollValues.scrollHeight - this.scrollValues.clientHeight > this.scrollValues.scrollTop);

        this.topShadow?.style.setProperty('opacity', this.scrollValues.scrollTop > 0 ? '1' : '0');
        this.bottomShadow?.style.setProperty('opacity', showBottomShadow ? '1' : '0');
    }

    updateScrollToFocus() {
        let focusCoord = this.props.focusedIndex && this.rowOffsets[this.props.focusedIndex] || 0;
        let rowHeight = this.props.focusedIndex && this.rowHeights[this.props.focusedIndex] || 0;
        let topElementCoord = this.scrollValues.scrollTop - rowHeight;
        let bottomElementCoord = this.scrollValues.scrollTop + this.scrollValues.clientHeight;

        if (focusCoord < topElementCoord || bottomElementCoord < focusCoord) {
            this.scrollBars.scrollTop(focusCoord - this.scrollValues.clientHeight / 2 + rowHeight / 2);
            this.scrollValues = this.scrollBars.getValues();
        }
    }

    componentDidMount() {
        if (process.env.JEST_WORKER_ID) {
            return;
        }

        this.updateRowHeights();
        this.handleUpdateScroll();
        this.updateScrollToFocus();
    }

    componentDidUpdate(prevProps: VirtualListProps) {
        this.updateRowHeights();
        this.handleUpdateScroll();
        if (this.props.focusedIndex !== prevProps.focusedIndex) {
            this.updateScrollToFocus();
        }
    }

    updateRefs(scrollBars: ScrollBars) {
        this.scrollBars = scrollBars;
        const root = scrollbars && ReactDOM.findDOMNode(scrollBars) as Element;
        this.container1 = root && root?.getElementsByClassName(css.container1)[0];
        this.container2 = root && root?.getElementsByClassName(css.container2)[0] as HTMLElement;
        this.container3 = root && root?.getElementsByClassName(css.container3)[0];
    }

    renderRows() {
        const topIndex = this.props.value?.topIndex || 0;
        const topY = this.rowOffsets[topIndex] || 0;

        return <div role={ this.props.role } className={ css.container3 } style={ { marginTop: topY } }>
            { this.props.rows }
        </div>;
    }

    updateRowHeights() {
        const nodes = this.container3?.children;
        const topIndex = this.props.value?.topIndex || 0;
        for (let n = 0; n < nodes.length; n++) {
            this.rowHeights[topIndex + n] = nodes[n].getBoundingClientRect().height;
        }
        const measured = this.rowHeights.filter(h => !!h);
        const averageHeight = measured.length > 0
            ? measured.reduce((sum, next) => sum + next, 0) / measured.length
            : 1;

        let lastOffset = 0;
        this.rowOffsets = [];
        for (let n = 0; n < this.props.rowsCount; n++) {
            this.rowOffsets[n] = lastOffset;
            let rowHeight = this.rowHeights[n];
            if (rowHeight == null) {
                rowHeight = averageHeight;
            }
            lastOffset += rowHeight;
        }

        let estimatedHeight = lastOffset;

        if (this.estimatedHeight != estimatedHeight) {
            this.container2.style.setProperty('min-height', `${estimatedHeight}px`);
        }

        this.estimatedHeight = estimatedHeight;
    }

    renderView({ style, ...props }: { style: {}, props: any }) {
        return (
            <div
                className={ css.container1 }
                style={ { ...style } }
                { ...props }
            />
        );
    }

    render() {
        return (
            <div className={ cx(css.wrapper, this.props.cx) } {...this.props.rawProps} >
                <ScrollBars
                    key='s'
                    autoHeight
                    className={ css.body }
                    onScroll={ this.handleUpdateScroll }
                    hideTracksWhenNotNeeded
                    renderView={ this.renderView.bind(this) }
                    ref={ el => this.updateRefs(el) }
                    autoHeightMax={ 100500 }
                    renderThumbHorizontal={ () => <div className='uui-thumb-horizontal'/> }
                    renderThumbVertical={ () => <div className='uui-thumb-vertical'/> }
                >
                    <div className={ css.container2 } style={ { minHeight: this.estimatedHeight } }>
                        { this.renderRows() }
                    </div>
                </ScrollBars>
                <div key='st' className='uui-scroll-shadow-top' ref={ el => this.topShadow = el } />
                <div key='sb' className='uui-scroll-shadow-bottom' ref={ el => this.bottomShadow = el } />
            </div>
        );
    }
}