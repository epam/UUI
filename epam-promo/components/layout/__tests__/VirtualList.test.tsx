import React from 'react';
import { VirtualList } from '../VirtualList';
import renderer from 'react-test-renderer';

describe('VirtualList', () => {
    it('should be rendered correctly', () => {
        const tree = renderer.create(<VirtualList rows={null} value={null} onValueChange={() => {}} />).toJSON();
        expect(tree).toMatchSnapshot();
    });
});
