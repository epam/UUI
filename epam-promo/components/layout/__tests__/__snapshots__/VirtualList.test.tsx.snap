// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VirtualList should be rendered correctly 1`] = `
<div
  className="wrapper root"
>
  <div
    className="body"
    style={
      Object {
        "height": "auto",
        "maxHeight": 100500,
        "minHeight": 0,
        "overflow": "hidden",
        "position": "relative",
        "width": "100%",
      }
    }
  >
    <div
      className="container1"
      style={
        Object {
          "WebkitOverflowScrolling": "touch",
          "bottom": undefined,
          "left": undefined,
          "marginBottom": 0,
          "marginRight": 0,
          "maxHeight": 100500,
          "minHeight": 0,
          "overflow": "scroll",
          "position": "relative",
          "right": undefined,
          "top": undefined,
        }
      }
    >
      <div
        className="container2"
        style={
          Object {
            "minHeight": null,
          }
        }
      >
        <div
          className="container3"
          role="listbox"
          style={
            Object {
              "marginTop": 0,
            }
          }
        />
      </div>
    </div>
    <div
      style={
        Object {
          "borderRadius": 3,
          "bottom": 2,
          "display": "none",
          "height": 6,
          "left": 2,
          "position": "absolute",
          "right": 2,
        }
      }
    >
      <div
        className="uui-thumb-horizontal"
      />
    </div>
    <div
      style={
        Object {
          "borderRadius": 3,
          "bottom": 2,
          "display": "none",
          "position": "absolute",
          "right": 2,
          "top": 2,
          "width": 6,
        }
      }
    >
      <div
        className="uui-thumb-vertical"
      />
    </div>
  </div>
  <div
    className="uui-scroll-shadow-top"
  />
  <div
    className="uui-scroll-shadow-bottom"
  />
</div>
`;
